lib/KBox/src/tasks/AutoPilotTask.cpp

```diff
@@ -0,0 +1,132 @@
+  in the Software without restriction, including without limitation the rights
+  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+  copies of the Software, and to permit persons to whom the Software is
+  furnished to do so, subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be included in
+  all copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  THE SOFTWARE.
+*/
+
+/* AutoPilot functionality can exist at several levels of complexity.  In this initial iteration, the design
+ focuses on heading the boat on a defined course that is automatically set when AP "Heading" Mode is activated
+ Later on we could introduce an additional mode when a course to a waypoint is provided by a NMEA-connected device.
+ "Dodge" mode suspends rudder actions temporarily to manually steer the boat around
+ an object or to manually put it back on course.
+
+ The course calculation approach below is a simplified and adapted version of Robert Huitema's excellent FreeboardPLC_v1_2 code. RIGM
+ */
+
+#include "AutoPilotTask.h"
+#include "../pages/NavigationPage.h"
+#include "KBoxDebug.h"
+#include "../drivers/board.h"
+#include "KMessage.h"
+#include <NMEA2000.h>
+#include <N2kMessages.h>
+
+AutoPilotTask::AutoPilotTask() : Task("AutoPilotTask"),
+  apCurrentHeading(0), apTargetHeading(0), apTargetRudderPosition(0),
+  headingPID(&apCurrentHeading, &apTargetRudderPosition, &apTargetHeading, (double)P_Param, (double)I_Param, (double)D_Param, (int) REVERSE)
+{
+
+}
+
+void AutoPilotTask::processMessage(const KMessage &message) {
+  message.accept(*this);
+}
+
+void AutoPilotTask::visit(const NAVMessage &nav) {
+  apCurrentHeading = nav.getCurrentHeading();
+  apTargetHeading = nav.getTargetHeading();
+  navMode = nav.getApMode();
+  navDodgeMode = nav.getApDodgeMode();
+}
+
+void AutoPilotTask::visit(const RUDMessage &rm) {
+  apRudderSensorPosition = rm.getRawRudderAngle();
+}
+
+void AutoPilotTask::setup() {
+  sameLastDirection=true;
+  apTargetRudderPosition = MAXRUDDERSWING / 2; //centred.  For programming purposes rudder is assumed to move through 66 degrees lock to lock
+  headingPID.SetMode(AUTOMATIC);
+  headingPID.SetOutputLimits(0.0, MAXRUDDERSWING); //output limits  0 = full starboard rudder (trailing edge of rudder to the right, bow moves to right)
+  headingPID.SetSampleTime(AUTOPILOT_SAMPLE_TIME);
+  DEBUG("Init autopilot complete");
+}
+
+void AutoPilotTask::loop() {
+  headingPID.Compute();
```

Comment:
You should know from testing the AP code on KBOX that it works!  That's the magic of the math approach that Robert Huitema developed.  Circular compass math is converted to to linear math by adding 720 and defining the heading data that feeds the PID function as targetheading plus offcourse.  If you set DEBUG on the numbers being fed into the PID you'll see how this works.  In effect, the inputs are already "preprocessed" as you suggest.

Have taken a brief closer look at moving the math functions from the current Nav page into the AP task.  That's going to take more time than I have available for the next few days.  I'll look again later in the week.  I'm not entirely sure at this point whether the practical benefits of doing so are worth the effort, particularly as I don't think we can entirely eliminate the math in the Nav page in any event.